v2: Reauthenticate failed with context cancelled

I'm trying to fix reauth in my fleeting plugin, because seems AllowReauth do not work, as i got:

```
Feb 19 09:38:57 runner-manager.ci.cloud gitlab-runner[882]: 2024-02-19T09:38:57.199Z [ERROR] plugin.fleeting-plugin-openstack: Failed to delete instance: cloud=runner
Feb 19 09:38:57 runner-manager.ci.cloud gitlab-runner[882]:   err=
Feb 19 09:38:57 runner-manager.ci.cloud gitlab-runner[882]:   | Unable to re-authenticate: Expected HTTP response code [202 204] when accessing [DELETE https://ci.cloud/compute/v2.1/servers/6442d29d-af4f-4d53-ae1a-3054e9ff3428], but got 401 instead
Feb 19 09:38:57 runner-manager.ci.cloud gitlab-runner[882]:   | {"error": {"code": 401, "title": "Unauthorized", "message": "The request you have made requires authentication."}}: Post "https://ci.cloud/identity/v3/auth/tokens": context canceled
Feb 19 09:38:57 runner-manager.ci.cloud gitlab-runner[882]:    id=6442d29d-af4f-4d53-ae1a-3054e9ff3428 name=podman-runners timestamp=2024-02-19T09:38:57.199Z
```

When i tried to do:
```golang
cli.Context = nil // ensure there no context!
err := cli.Reauthenticate(cli.TokenID())
```
The result error:
```
Re-authenticate failed: err="Post \"https://ci.cloud/identity/v3/auth/tokens\": context canceled"
```

--- a/internal/acceptance/openstack/client_test.go
+++ b/internal/acceptance/openstack/client_test.go
@@ -4,6 +4,7 @@
 package openstack
 
 import (
+	"context"
 	"os"
 	"testing"
 	"time"
@@ -140,7 +141,7 @@ func TestReauth(t *testing.T) {
 	time.Sleep(1 * time.Second)
 	t.Logf("Attempting to reauthenticate")
 
-	err = provider.ReauthFunc()
+	err = provider.ReauthFunc(context.TODO())
 	if err != nil {
 		t.Fatalf("Unable to reauthenticate: %v", err)
 	}

--- a/openstack/client.go
+++ b/openstack/client.go
@@ -170,7 +170,7 @@ func v2auth(ctx context.Context, client *gophercloud.ProviderClient, endpoint st
 		tac.SetTokenAndAuthResult(nil)
 		tao := options
 		tao.AllowReauth = false
-		client.ReauthFunc = func() error {
+		client.ReauthFunc = func(ctx context.Context) error {
 			err := v2auth(ctx, &tac, endpoint, tao, eo)
 			if err != nil {
 				return err
@@ -293,7 +293,7 @@ func v3auth(ctx context.Context, client *gophercloud.ProviderClient, endpoint st
 		default:
 			tao = opts
 		}
-		client.ReauthFunc = func() error {
+		client.ReauthFunc = func(ctx context.Context) error {
 			err := v3auth(ctx, &tac, endpoint, tao, eo)
 			if err != nil {
 				return err

--- a/provider_client.go
+++ b/provider_client.go
@@ -83,7 +83,7 @@ type ProviderClient struct {
 	// ReauthFunc is the function used to re-authenticate the user if the request
 	// fails with a 401 HTTP response code. This a needed because there may be multiple
 	// authentication functions for different Identity service versions.
-	ReauthFunc func() error
+	ReauthFunc func(context.Context) error
 
 	// Throwaway determines whether if this client is a throw-away client. It's a copy of user's provider client
 	// with the token and reauth func zeroed. Such client can be used to perform reauthorization.
@@ -273,12 +273,14 @@ func (client *ProviderClient) SetThrowaway(v bool) {
 // reauthenticated in the meantime. If no previous token is known, an empty
 // string should be passed instead to force unconditional reauthentication.
 func (client *ProviderClient) Reauthenticate(previousToken string) error {
+	ctx := context.TODO()
+
 	if client.ReauthFunc == nil {
 		return nil
 	}
 
 	if client.reauthmut == nil {
-		return client.ReauthFunc()
+		return client.ReauthFunc(ctx)
 	}
 
 	future := newReauthFuture()
@@ -299,7 +301,7 @@ func (client *ProviderClient) Reauthenticate(previousToken string) error {
 	// Perform the actual reauthentication.
 	var err error
 	if previousToken == "" || client.TokenID == previousToken {
-		err = client.ReauthFunc()
+		err = client.ReauthFunc(ctx)
 	} else {
 		err = nil
 	}

--- a/testing/provider_client_test.go
+++ b/testing/provider_client_test.go
@@ -66,7 +66,7 @@ func TestConcurrentReauth(t *testing.T) {
 	p := new(gophercloud.ProviderClient)
 	p.UseTokenLock()
 	p.SetToken(prereauthTok)
-	p.ReauthFunc = func() error {
+	p.ReauthFunc = func(_ context.Context) error {
 		p.SetThrowaway(true)
 		time.Sleep(1 * time.Second)
 		p.AuthenticatedHeaders()
@@ -154,7 +154,7 @@ func TestReauthEndLoop(t *testing.T) {
 	p := new(gophercloud.ProviderClient)
 	p.UseTokenLock()
 	p.SetToken(client.TokenID)
-	p.ReauthFunc = func() error {
+	p.ReauthFunc = func(_ context.Context) error {
 		info.mut.Lock()
 		defer info.mut.Unlock()
 
@@ -237,7 +237,7 @@ func TestRequestThatCameDuringReauthWaitsUntilItIsCompleted(t *testing.T) {
 	p := new(gophercloud.ProviderClient)
 	p.UseTokenLock()
 	p.SetToken(prereauthTok)
-	p.ReauthFunc = func() error {
+	p.ReauthFunc = func(_ context.Context) error {
 		info.mut.RLock()
 		if info.numreauths == 0 {
 			info.mut.RUnlock()
@@ -331,7 +331,7 @@ func TestRequestReauthsAtMostOnce(t *testing.T) {
 	p := new(gophercloud.ProviderClient)
 	p.UseTokenLock()
 	p.SetToken(client.TokenID)
-	p.ReauthFunc = func() error {
+	p.ReauthFunc = func(_ context.Context) error {
 		reauthCounterMutex.Lock()
 		reauthCounter++
 		reauthCounterMutex.Unlock()

